clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -535,16 +532,13 @@ CreateSYCLKernelBody(Sema &S, FunctionDecl *KernelCallerFunc, DeclContext *DC) {
 
         // __init needs four parameter
         auto ParamItr = InitMethod->param_begin();
+
         // kernel_parameters
-        llvm::SmallVector<Expr *, NumParams> ParamStmts;
-        ParamStmts.push_back(getExprForPointer(
-            S, (*(ParamItr++))->getOriginalType(), ParamDREs[0]));
-        ParamStmts.push_back(getExprForRangeOrOffset(
-            S, ((*ParamItr++))->getOriginalType(), ParamDREs[1]));
-        ParamStmts.push_back(getExprForRangeOrOffset(
-            S, ((*ParamItr++))->getOriginalType(), ParamDREs[2]));
-        ParamStmts.push_back(getExprForRangeOrOffset(
-            S, ((*ParamItr++))->getOriginalType(), ParamDREs[3]));
+        llvm::SmallVector<Expr *, 4> ParamStmts;
+        for (size_t I = 0; I < NumParams; ++I) {
+          ParamStmts.push_back(getExprForKernelParameter(
+              S, (*(ParamItr++))->getOriginalType(), ParamDREs[I]));
+        }
         // [kenrel_obj or wrapper object].accessor.__init(_ValueType*,
```

Comment:
Since you are visiting the area :-) what about `kernel_obj`?

Reply:
No problem. Done.